### PR TITLE
added new classes and functions

### PR DESCRIPTION
**Pull request by vKxni**
The following PR shows the following (new) classes and functions 

### All new `classes`: 

- ArrayRange (https://hexdocs.pm/elixir/1.13/Range.html)
Represents a sequence of ascending or descending integers with a common difference called step.

```ts
const range1 = new ArrayRange(1, 10);
console.log([...range1]); // => [1, 2, 3 ... 10]
```

- Kernel (https://hexdocs.pm/elixir/1.14/Kernel.html)
The `Kernel` class provides a collection of utility functions similar to those found in the Elixir standard library.

```ts
const kernel = new Kernel();

console.log(kernel.binaryPart("hello", 1, 3)); // => "ell"
console.log(kernel.bitSize("110101")); // => 6
console.log(kernel.byteSize("110101")); // => 1
console.log(kernel.ceil(3.14)); // => 4
console.log(kernel.div(10, 3)); // => 3

const tuple = [1, "hello", true];
console.log(kernel.elem(tuple, 1)); // => "hello"
console.log(kernel.floor(3.14)); // => 3


const list = [10, 20, 30];
console.log(kernel.hd(list)); // => 10

console.log(kernel.in(3, [1, 2, 3])); // => true

console.log(kernel.isAtom("hello")); // => true

console.log(kernel.isBinary("hello")); // => true

console.log(kernel.isBoolean(true)); // => true

console.log(kernel.isException(new Error("Some error"))); // => true

console.log(kernel.isFloat(3.14)); // => true

console.log(kernel.isInteger(42)); // => true

console.log(kernel.isList([1, 2, 3])); // => true

console.log(kernel.isMap({ name: "John", age: 30 })); // => true

console.log(kernel.isNil(null)); // => true

console.log(kernel.isNumber(42)); // => true

console.log(kernel.isPid({ pid: "process_id@node" })); // => true

console.log(kernel.isPort({ port: "port_id@node" })); // => true

console.log(kernel.isReference({ ref: "ref_id@node" })); // => true

console.log(kernel.isTuple([1, 2, 3])); // => true

console.log(kernel.length([1, 2, 3])); // => 3

console.log(kernel.mapSize({ name: "John", age: 30 })); // => 2

console.log(kernel.not(true)); // => false

console.log(kernel.or(true, false)); // => true

console.log(kernel.rem(10, 3)); // => 1

console.log(kernel.round(3.14)); // => 3

console.log(kernel.tl([1, 2, 3])); // => [2, 3]

console.log(kernel.trunc(3.14)); // => 3

console.log(kernel.tupleSize([1, 2, 3])); // => 3
```

- MapAccess (https://hexdocs.pm/elixir/1.12.3/Access.html)
Access behavior implementation for keyword lists. Or a Key-based access to data structures.

```ts
const mapData: Map<unknown, unknown> = new Map();
mapData.set('name', 'John');
mapData.set(42, 'answer');
mapData.set('email', 'john@example.com');

const mapAccess = new MapAccess(mapData);

console.log(mapAccess.get('name')); // John
console.log(mapAccess.get(42)); // answer
console.log(mapAccess.get({ key: 'email' })); // john@example.com
console.log(mapAccess.get('gender')); // null
```

- MapSet (https://hexdocs.pm/elixir/1.13/MapSet.html)
A class representing a Set data structure.
A set is a data structure that can contain unique elements of any kind, without any particular order. MapSet is the "go to" set data structure in Elixir.
A set can be constructed using `MapSet.new/0`
```ts
const mapSet = MapSet.new<number>([2]);
mapSet.put(4);
mapSet.put(0);
console.log(mapSet.values()); // => [2, 4, 0]

mapSet.put(4); // Should not add duplicate element
console.log(mapSet.values()); // => [2, 4, 0]

console.log(mapSet.has(4)); // => true
console.log(mapSet.has(5)); // => false

mapSet.remove(4);
console.log(mapSet.values()); // => [2, 0]

const otherSet = MapSet.new<number>([2, 4]);
console.log(mapSet.intersection(otherSet).values()); // => [2]
console.log(mapSet.difference(otherSet).values()); // => [0]
console.log(mapSet.union(otherSet).values()); // => [2, 0, 4]

const subset = MapSet.new<number>([2]);
console.log(subset.isSubset(mapSet)); // => true
```

- OptionParser (https://hexdocs.pm/elixir/1.13.4/OptionParser.html)
Functions for parsing command line arguments.

```ts
const argv: string[] = ["--debug", "--output", "output.txt"];
const opts: Options = { debug: "boolean", => "string" };
const parsedOptions = OptionParser.parse(argv, opts);
console.log(parsedOptions); // => { debug: true, => 'output.txt' }

const str = "--verbose --input input.txt";
const argvFromString = OptionParser.split(str);
console.log(argvFromString); // => [ '--verbose', '--input', 'input.txt' ]

const options = { verbose: true, input: "input.txt" };
const argvFromOptions = OptionParser.toArgv(options);
console.log(argvFromOptions); // => [ '--verbose', '--input', 'input.txt' ]
```

- FileInfoConverter (https://hexdocs.pm/elixir/1.13/File.html)
This module contains functions to manipulate/show statistics of files.

```ts
const elixirFileInfo = {
    size: 1024,
    type: 'regular',
    access: 'read_write',
    atime: 1669820525000,
    mtime: '2023-08-01T12:34:56.789Z',
    ctime: '2023-08-01T12:34:56.789Z',
    mode: 0o755,
    links: 1,
    major_device: 0,
    minor_device: 0,
    inode: 12345,
    uid: 1000,
    gid: 1000,
};

const tsFileStat = FileInfoConverter.fromRecord(elixirFileInfo);
console.log(tsFileStat);

const tsFileInfo = FileInfoConverter.toRecord(tsFileStat);
console.log(tsFileInfo);
{
  size: 1024,
  type: 'regular',
  access: 'read_write',
  atime: 1669820525000,
  mtime: 2023-08-01T12:34:56.789Z,
  ctime: 2023-08-01T12:34:56.789Z,
  mode: 493,
  links: 1,
  major_device: 0,
  minor_device: 0,
  inode: 12345,
  uid: 1000,
  gid: 1000
}
{
  size: 1024,
  type: 'regular',
  access: 'read_write',
  atime: 1669820525000,
  mtime: '2023-08-01T12:34:56.789Z',
  ctime: '2023-08-01T12:34:56.789Z',
  mode: 493,
  links: 1,
  major_device: 0,
  minor_device: 0,
  inode: 12345,
  uid: 1000,
  gid: 1000
}
```

### All new `functions`

- enumMap
Applies the provided callback function to each element of the input array and returns a new array with the results.

```ts
const array: number[] = [1, 2, 3];
const out: number[] = enumMap(array, (x) => x * 2)
console.log(out); // => [2, 4, 6]
```

- keywordList
Converts a keyword list in array format to object format.

```ts
const options: KeywordList = [
    ["exit_on_close", true],
    ["active", "once"],
    ["packet_size", 1024],
];

// Converting the keyword list to an object
const optionsObject = keywordListToObject(options);
console.log(optionsObject); // => { exit_on_close: true, active: 'once', packet_size: 1024 }

// Accessing values from the keyword list object
console.log(optionsObject.exit_on_close); // => true
console.log(optionsObject.active); // => 'once'
console.log(optionsObject.packet_size); // => 1024

// Modifying values in the keyword list object
optionsObject.active = 'continuous';
console.log(optionsObject.active); // => 'continuous'

// Converting the keyword list object back to an array
const updatedOptionsList = keywordObjectToList(optionsObject);
console.log(updatedOptionsList); // => [ [ 'exit_on_close', true ], [ 'active', 'continuous' ], [ 'packet_size', 1024 ] ]
```

- mapObjectValues
Converts a keyword list in array format to object format.

```ts
const inputMap: Record<string, number> = { "a": 1, "b": 2 };
const outputMap: Record<string, number> = mapObjectValues(inputMap, (k, v) => v * 2);
console.log(outputMap); // { a: 2, b: 4 }
```

vKxni - 08/04/2023

### CODE REVIEW

File: src/dev/classes/ArrayRange.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
+/**
+ * Represents a sequence of ascending or descending integers with a common difference called step.
+ * Ranges are always inclusive.
+ * 
+ * @see https://hexdocs.pm/elixir/1.13/Range.html
+ */
+class ArrayRange {
+    private _first: number;
```
Comment (adambrangenberg) [inline src/dev/classes/ArrayRange.ts:8]: Use Hashtags instead of Underscores for real private :)

File: src/dev/classes/ArrayRange.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    toArray(): number[] {
+        return [...this.generateRange()];
+    }
+
+    /**
+     * Returns a generator that yields all the values in the range.
+     * @returns A generator yielding the elements of the range.
+     */
+    private *generateRange(): Generator<number, void, unknown> {
```
Comment (adambrangenberg) [inline src/dev/classes/ArrayRange.ts:65]: why *?

**Comment by vKxni** [comment]
I noticed that the example at `mapObjectValues` somehow was the same as the one from `keywordList` lol mb

File: src/dev/classes/FileInfoConverter.ts [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        mode: number;
+        links: number;
+        major_device: number;
+        minor_device: number;
+        inode: number;
+        uid: number;
+        gid: number;
+    }): FileStat {
+        return {
+            size: fileInfo.size,
```
Comment (adambrangenberg) [inline src/dev/classes/FileInfoConverter.ts:99]: Consider just reassigning the values or something like that?

File: src/dev/classes/FileInfoConverter.ts [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+```
+ */
+class FileInfoConverter {
+    /**
+     * Converts a :file_info record into a File.Stat.
+     * @param fileInfo The :file_info record to convert.
+     * @returns File.Stat object.
+     * 
+     */
+    static fromRecord(fileInfo: {
```
Comment (adambrangenberg) [inline src/dev/classes/FileInfoConverter.ts:83]: why not FileStat type?

File: src/dev/classes/FileInfoConverter.ts [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+     * @param stat The File.Stat object to convert.
+     * @returns :file_info record.
+     */
+    static toRecord(stat: FileStat): any {
+        return {
+            size: stat.size,
+            type: stat.type,
+            access: stat.access,
+            atime: FileInfoConverter.formatDateTime(stat.atime),
+            mtime: FileInfoConverter.formatDateTime(stat.mtime).toString(), // Convert DateTime to string
```
Comment (adambrangenberg) [inline src/dev/classes/FileInfoConverter.ts:126]: same here

File: src/dev/classes/MapAccess.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+ * MapAccess implements the Access interface for dynamically accessing keys in a Map data structure.
+ * @template K The type of keys in the Map.
+ * @template V The type of values in the Map.
+ */
+class MapAccess<K, V> implements Access<V> {
+    /**
+     * The Map object to be accessed.
+     * @private
+     */
+    private _map: Map<K, V>;
```
Comment (adambrangenberg) [inline src/dev/classes/MapAccess.ts:64]: same here

File: src/dev/classes/MapSet.ts [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    console.log(mapSet.intersection(otherSet).values()); // => [2]
+    console.log(mapSet.difference(otherSet).values()); // => [0]
+    console.log(mapSet.union(otherSet).values()); // => [2, 0, 4]
+
+    const subset = MapSet.new<number>([2]);
+    console.log(subset.isSubset(mapSet)); // => true
+```
+*/
+class MapSet<T> {
+    private elements: Set<T>;
```
Comment (adambrangenberg) [inline src/dev/classes/MapSet.ts]: why others with _ and here not?

File: src/dev/classes/MapSet.ts [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+    /**
+     * Creates a new MapSet instance.
+     * @param elements An array of elements to initialize the set.
+     */
+    constructor(elements: T[] = []) {
```
Comment (adambrangenberg) [inline src/dev/classes/MapSet.ts:40]: Maybe make them parse a Set, considering some values may vanish?

File: src/dev/classes/OptionParser.ts [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+     * @param argv - The array of command line options and arguments.
+     * @param opts - The options configuration that specifies the type of each option.
+     * @returns The parsed options as a keyword list.
+     * @throws An Error if any invalid options are given.
+     */
+    static parse(argv: string[], opts: Options = {}): ParsedOptions {
+        const result: ParsedOptions = {};
+        let i = 0;
+
+        while (i < argv.length) {
```
Comment (adambrangenberg) [inline src/dev/classes/OptionParser.ts]: why no for loop?

File: src/dev/classes/OptionParser.ts [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    /**
+     * Similar to `parse`, but only parses the head of `argv`.
+     * As soon as it finds a non-switch, it stops parsing.
+     * @param argv - The array of command line options and arguments.
+     * @param opts - The options configuration that specifies the type of each option.
+     * @returns The parsed options as a keyword list.
+     */
+    static parseHead(argv: string[], opts: Options = {}): ParsedOptions {
+        const result: ParsedOptions = {};
+        let i = 0;
```
Comment (adambrangenberg) [inline src/dev/classes/OptionParser.ts]: i is never used?

File: src/dev/enumMap.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+ * 
+ * ```ts
+    const array: number[] = [1, 2, 3];
+    const out: number[] = enumMap(array, (x) => x * 2)
+    console.log(out); // [2, 4, 6]
+```
+ */
+function enumMap<T, R>(arr: T[], callback: (value: T) => R): R[] {
+    const result: R[] = [];
+    const length = arr.length;
```
Comment (adambrangenberg) [inline src/dev/enumMap.ts:15]: whyyy

**Review by adambrangenberg — COMMENTED** [review]
deez nuts hihi, idk what to say here

File: src/dev/classes/OptionParser.ts (hunk 10 above)
Comment (vKxni) [reply src/dev/classes/OptionParser.ts]: it was, right below. but now moved it directly into the loop to be for(let i ...) lol

**Comment by vKxni** [comment]
> deez nuts hihi, idk what to say here

Thenks

**Pull request merged**